src/widgets/index/index.ts

```diff
@@ -264,16 +264,51 @@ const index = (props: IndexProps): Index => {
           _uiState: localUiState,
         });
 
+        localWidgets.forEach(widget => {
+          if (widget.getWidgetRenderState) {
+            const widgetRenderState = widget.getWidgetRenderState(
+              localInstantSearchInstance!.renderState[this.getIndexId()] || {},
+              {
```

Comment:
You could store these parameters in variables then re-use them in the `init`, `render` and `getWidgetRenderState` calls.

It would simplify the code (no more repeated code) and make it more readable imho:

```js
const renderState = instantSearchInstance.renderState[this.getIndexId()] || {};

const initOptions = {
    uiState,
    helper: this.getHelper()!,
    parent: this,
    instantSearchInstance,
    state: helper!.state,
    renderState: instantSearchInstance.renderState,
    templatesConfig: instantSearchInstance.templatesConfig,
    createURL,
    results: undefined,
    scopedResults: [],
    searchMetadata: {
        isSearchStalled: instantSearchInstance._isSearchStalled,
    },
};

const widgetRenderState = widget.getWidgetRenderState(renderState, initOptions);
```

Reply:
The params values are not the same during `init` and `render` so we cannot share them between these two methods.

Reusing the same params between `getWidgetRenderState` and `init|render` might leads to bug because they could expose stale values.

Reply:
Not sharing between `init` and `render` but between `getWidgetRenderState`/`init` and then `getWidgetRenderState`/`render`.

For the stale value problem, the 2 calls are synchronous and the only thing that is updated between them is `renderState`, it could be passed separately from the option object to avoid that.

Reply:
There's a lot of properties going on in that function call, but I think similar code "twice" is still okay.

Reply:
Yeah, since our internal state is based on mutation, it's too dangerous to reuse the same reference in the two function calls. I just tried and the current tests fail. It's safer to duplicate the code here.

Reply:
Ok, I find this very concerning that it is "dangerous" to do this. But this is more relevant to bigger problems on how we manage the state.